app/src/org/commcare/dalvik/geo/HereFunctionHandler.java

```diff
@@ -0,0 +1,145 @@
+import android.os.Bundle;
+import android.os.Looper;
+import android.support.v4.content.ContextCompat;
+import android.util.Log;
+
+import org.commcare.dalvik.activities.EntitySelectActivity;
+import org.commcare.dalvik.application.CommCareApplication;
+import org.javarosa.core.model.condition.EvaluationContext;
+import org.javarosa.core.model.condition.IFunctionHandler;
+import org.javarosa.core.model.data.GeoPointData;
+import org.odk.collect.android.utilities.GeoUtils;
+
+import java.util.Set;
+import java.util.Vector;
+
+/**
+ * HereFunctionHandler is a class that can be used to evaluate the here() XPath expression.
+ * In any context where here() needs to be evaluated, create a HereFunctionHandler and
+ * call addFunctionHandler to add it to the XPath evaluator.
+ *
+ * In addition, an EntitySelectActivity can register itself to be refreshed whenever
+ * a new value of here() is obtained (whenever the location changes).
+ *
+ * No locations are requested if here() is never evaluated.
+ *
+ * @author Forest Tong
+ */
+public class HereFunctionHandler implements IFunctionHandler, LocationListener {
+    public static final String HERE_NAME = "here";
+    private GeoPointData location;
+
+    private LocationManager mLocationManager;
+
+    // This flag keeps track of whether the location manager needs to be set up and subscribed to.
+    private boolean locationUpdatesRequested = false;
+
+    private final Context context = CommCareApplication._().getApplicationContext();
+    // If there are more general uses for HereFunctionHandler, the type of this field can be
+    // generalized to a listener interface.
+    private EntitySelectActivity entitySelectActivity;
+
+    public HereFunctionHandler() {}
+
+    public void registerEntitySelectActivity(EntitySelectActivity entitySelectActivity) {
+        this.entitySelectActivity = entitySelectActivity;
+    }
+
+    // Setting up and subscribing to the location manager is delayed until here() is actually
+    // evaluated.
+    public Object eval(Object[] args, EvaluationContext ec) {
+        if(!locationUpdatesRequested) requestLocationUpdates();
+        if(location == null) return "";
+        return location.getDisplayText();
+    }
+
+    private void requestLocationUpdates() {
+        mLocationManager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
+        Set<String> mProviders = GeoUtils.evaluateProviders(mLocationManager);
+
+        for (String provider : mProviders) {
+            if ((provider.equals(LocationManager.GPS_PROVIDER) && ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) ||
+                    (provider.equals(LocationManager.NETWORK_PROVIDER) && ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED)) {
+                // This is non-null if the calling activity pauses.
```

Comment:
Basically what I meant to point out is that if `eval` is never called, then the location is never requested. But I already mentioned this at the top, and I guess it's pretty self-explanatory, so I'll take out the comment.